crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -45,6 +50,166 @@ pub trait WorldQuery {
     type State: FetchState;
 }
 
+/// Types that implement this trait are responsible for fetching query items from tables or
+/// archetypes.
+///
+/// Every type that implements [`WorldQuery`] have their associated [`WorldQuery::Fetch`] and
+/// [`WorldQuery::State`] types that are essential for fetching component data. If you want to
+/// implement a custom query type, you'll need to implement [`Fetch`] and [`FetchState`] for
+/// those associated types.
+///
+/// You may want to implement a custom query for the following reasons:
```

Comment:
There's another reason to add a new WorldQuery impl: you want a fundamentally new type of behavior. I ran into this in #2116, where I was working on implementing `EventReader` and `EventWriter` for components. This won't work with the derive though, and I'm not quite sure how to frame.